### PR TITLE
W3CLogger Improvement

### PR DESCRIPTION
**Pull request by wcontayon**
**W3CLogger Improvement**

**PR Description**
- Write more content directly to StreamWriter in W3CLogger

Addresses #33946

### CODE REVIEW

File: src/Middleware/HttpLogging/test/FileLoggerProcessorTests.cs [hunk 1]
```diff
@@ -285,5 +285,8 @@ private async Task WaitForRoll(string fileName)
                 await Task.Delay(100);
             }
         }
+
+        private string SetW3cFileName(string path, DateTimeOffset date, W3CLoggerOptions options, string suffix)
```
Comment (wtgodbe) [inline src/Middleware/HttpLogging/test/FileLoggerProcessorTests.cs:289]: `SetW3cFileName` -> `SetW3CFileName`

**Review by wtgodbe — CHANGES_REQUESTED** [review]

File: src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs [hunk 2]
```diff
@@ -118,5 +203,64 @@ internal override Task WriteMessageAsync(string message, StreamWriter streamWrit
...
+                try
+                {
+                    _messageQueue.Add(string.Join(W3CSepator, messages));
+                    return;
+                }
+                catch (InvalidOperationException) { }
+            }
+        }
+
+        internal override string FormatMessage(string message)
```
[the remark this answers is not in the document]
Comment (wcontayon) [reply src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs]: I got it 🙂 . I'll push a refacto of that part.

File: src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs [hunk 3]
```diff
@@ -35,75 +35,82 @@ public override async Task OnFirstWrite(StreamWriter streamWriter, CancellationT
...
             if (_loggingFields.HasFlag(W3CLoggingFields.Referer))
             {
-                await WriteField(" cs(Referer)", streamWriter, cancellationToken);
+                await streamWriter.WriteAsync(" cs(Referer)".AsMemory(), cancellationToken);
             }
+
+            await streamWriter.FlushAsync();
         }
 
         internal async Task WriteField(string messageField, StreamWriter streamWriter, CancellationToken cancellationToken)
```
Comment (wtgodbe) [inline src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs:116]: Do we still need this?

**Review by wtgodbe — CHANGES_REQUESTED** [review]
Looks like tests are failing because the Separator isn't being stripped out - this should be good once that's fixed (and my other comment on WriteField)

File: src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs [hunk 4]
```diff
@@ -113,10 +205,77 @@ private string GetFieldsDirective()
...
         // Extensibility point for tests
         internal virtual void OnWrite(string message) { }
+
+        public void EnqueueMessage(string[] messages)
+        {
+            if (!_messageQueue.IsAddingCompleted)
+            {
+                try
+                {
+                    _messageQueue.Add(string.Join(W3CSeparator, messages));
```
Comment (wtgodbe) [inline src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs:226]: Doing string.Join here & string.Split later means we aren't necessarily saving any allocations versus using the StringBuilder - we should try to think of a way to eliminate this and just write each array element directly to the streamWriter, though admittedly that might be complicated with having to enqueue them into the collection of strings.

Comment (Tratcher) [reply src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs:226]: The queue itself will need to change.

File: src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs (hunk 3 above)
Comment (wcontayon) [reply src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs:116]: No, I'll remove it. I was trying to write directly the fields in the streamwriter with this method, but I got a timeout with the test 🙂 . We can try a PR on that with a good refacto.

File: src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs (hunk 4 above)
Comment (wtgodbe) [reply src/Middleware/HttpLogging/src/W3CLoggerProcessor.cs:228]: @wcontayon still need to make this change in order to actually save allocations vs. what's currently in `main`

**Comment by davidfowl** [comment]
Did we profile this component? Using StreamWriter is concerning.

**Comment by wtgodbe** [comment]
> Did we profile this component? Using StreamWriter is concerning.

We did not, how can I go about doing so? What would you suggest using instead of StreamWriter?

**Comment by wcontayon** [comment]
Hi @wtgodbe, @Tratcher, hope you're doing well.
Sorry for the long absence, due to a long and complex mission 😅. Do we always need the improvement, on w3clogger ? if so can I re-submit a new PR with the changes ?
@Tratcher you suggest to change the queue for this improvement ?

**Pull request closed without merging**